# 159987629 User should be able to create articles

emmaadesile opened the pull request:
#### What does this PR do?
Implement the functionality that enables users to create articles

#### Description of Task to be completed?
- Create a simple to navigate Text editor where users
  - write and format text
  - add images
  - add links
  - add bullet points

#### How should this be manually tested?
- clone the app and run `npm install` to install app dependencies
- run `npm start`
- login at `localhost:300/login`
- navigate to `localhost:3000/create-articles`
- provide the title and body on the article page
- click on the publish button

#### Any background context you want to provide?
- After publishing an article, users need to redirected to a page that fetches that article but the feature is not ready yet

#### What are the relevant pivotal tracker stories?
[159987629](https://www.pivotaltracker.com/story/show/159987629)

#### Screenshots (if appropriate)
<img width="1440" alt="screen shot 2018-10-12 at 3 38 05 pm" src="https://user-images.githubusercontent.com/26192691/46876014-1ffc0a00-ce35-11e8-8974-45d60dc6d1d8.png">



#### Questions:

emmaadesile committed:
ft(create-artilces): users can create articles

- users can apply basic formatting to text

[Deliver 159987629]

emmaadesile committed:
ft(create-article): add support for uploading images

[Deliver 159987629]

emmaadesile committed:
fix merge conflicts from remote branch

emmaadesile committed:
ft(create-articles-test): write tests for create-articles endpoint

- write tests for article action
- write tests for article reducer
- wirte tests for article component

[Delivers 159987629]

emmaadesile committed:
nitpicks(create-article): fix merge conflicts

emmaadesile committed:
ft(create-articles): add validation to article creation

- user cannot create article without title
- user cannot create article without body

[Delivers 159987629]

emmaadesile committed:
ft(create-articles): add more tests

- test util function
- test article action and article reducer

[Delivers 159987629]

emekafredy reviewed (changes requested):
The .env file should always be ignored

emekafredy commented on src/actionTypes/article.js:
```diff
@@ -0,0 +1,3 @@
+export const CREATE_ARTICLE_SUCCESS = "CREATE_ARTICLE_SUCCESS";
```
You have 3 identical action types in ```src/actionTypes/article.js``` and  ```src/actionTypes/auth.js```
Any reason for this?

emmaadesile replied:
That was an oversight. I'll fix it now.

emekafredy commented on src/actions/article.js:
```diff
@@ -0,0 +1,51 @@
+    })
+    .then(response => {
+      if (response.data.status === "success") {
+        // const { slug } = response.data;
```
Always remove commented-out codes like in lines 39 and 40

emmaadesile replied:
okay

emekafredy commented on src/containers/CreateArticle/ArticleTitle.js:
```diff
@@ -0,0 +1,36 @@
+import { Editor } from "@tinymce/tinymce-react";
+import PropTypes from "prop-types";
+
+const apiKey = "0naso5vy7fw5eu2eqblor00o1c326x0hupermg4cli0rqoqn";
```
You should always values like on line 5 in the env.

johnkingzy replied:
@emmaadesile resolve this.

emmaadesile replied:
I have resolved this

emekafredy commented on src/containers/CreateArticle/CreateArticlePage.js:
```diff
@@ -0,0 +1,254 @@
+    try {
+      const formData = new FormData();
+      formData.append("file", blobInfo.blob());
+      formData.append("upload_preset", "asgjqcgx");
```
Your cloudinary keys and url (on lines 94 and 97)should be better set on the .env file

emmaadesile replied:
okay, will do that now

emekafredy commented on src/containers/CreateArticle/CreateArticlePage.js:
```diff
@@ -0,0 +1,254 @@
+  handleSubmit = () => {
+    const { title, body, error } = this.state;
+    const { createArticle } = this.props;
+    const description = stripHtml(body).slice(0, 200);
```
Nice work on generating this description from the body. On line 151, I think having 200 words (if I am correct on what I'm seeing) is a bit too much. In my opinion, the 30 first characters (or 25) from the body should be enough for the description.

emmaadesile replied:
It gets the first 200 characters in the text ie letters, spaces, etc.

CEOehis replied:
I think 200 characters is just fine.

emekafredy replied:
Oh! my bad. 200 characters is cool. I thought you were generating words, not characters.

CEOehis commented on src/actions/article.js:
```diff
@@ -0,0 +1,51 @@
+
+  dispatch(setArticleLoading(true));
+  return axios
+    .post("https://thor-ah-staging.herokuapp.com/api/articles", articleData, {
```
The Base API URL, `https://thor-ah-staging.herokuapp.com` should not be hard-coded. In order to allow you write your code your development environment you should extract this to a `process.env.REACT_APP_API` and add ```
REACT_APP_API='http://localhost:3000'
``` to your `.env`.

This way you can have dynamic base API URL's

johnkingzy commented on src/actions/article.js:
```diff
@@ -0,0 +1,51 @@
+
+dotenv.config();
+
+const setArticleSuccess = article => ({
```
This should follow the naming convention `createArticle` `createArticleSuccess` `createArticleLoading` `createArticleError`.

johnkingzy commented on src/containers/CreateArticle/CreateArticlePage.js:
```diff
@@ -0,0 +1,254 @@
+      formData.append("upload_preset", "asgjqcgx");
+      formData.append("api_key", "329369412183662");
+      const uploadUrl =
+        "https://api.cloudinary.com/v1_1/dn6fnuhxr/image/upload";
```
The Cloudinary URL shouldn't be hard-coded as well. you might want to add it to the `.env` file

johnkingzy commented on src/containers/CreateArticle/CreateArticlePage.js:
```diff
@@ -0,0 +1,254 @@
+import { Redirect } from "react-router-dom";
+import axios from "axios";
+import PropTypes from "prop-types";
+import createArticle from "../../actions/article";
```
You might want to leave a blank line and a comment between absolute import and relative imports

johnkingzy commented on src/containers/CreateArticle/CreateArticlePage.js:
```diff
@@ -0,0 +1,254 @@
+    if (title.trim() === "Title") {
+      this.setState({ title: "" });
+    }
+  };
```
I think the method `titlePlaceholderFocusInHandler` and `titlePlaceholderFocusOutHandler` seems to be implementing quite similar logic. I think you can have an identifier to detect which event is been called rather than having 2 separate methods. This also applies for the change Handler

emmaadesile replied:
Thanks for the feedback. I just figured out a way to do it and I have implemented it.

johnkingzy commented on src/containers/CreateArticle/TextEditor.js:
```diff
@@ -0,0 +1,61 @@
+import { Editor } from "@tinymce/tinymce-react";
+import PropTypes from "prop-types";
+
+const apiKey = "0naso5vy7fw5eu2eqblor00o1c326x0hupermg4cli0rqoqn";
```
This should be stored in the `.env` file

emmaadesile replied:
I have done that.

johnkingzy commented on src/reducers/article.js:
```diff
@@ -0,0 +1,30 @@
+        article: action.payload
+      };
+    case CREATE_ARTICLE_FAILURE:
+      return {
```
Let's try to stick with the naming convention, I will suggest you change this to `CREATE_ARTICLE_ERROR` instead

emmaadesile replied:
I have implemented the feedback

johnkingzy reviewed (changes requested):
Nice one! Fix conflicts as well

emmaadesile committed:
nitpicks(create-articles): add api keys to .env file

[Delivers 159987629]

emmaadesile committed:
nitpicks(create-articles): implement feedback

- rename CREATE_ARTICLE_FAILURE to CREATE_ARTICLE_ERROR
- handle title placeholder focusin and focusout with `titlePlaceholderHandler` method
- handle body placeholder focusin and focusout with `bodyPlaceholderHandler` method
- use Define Plugin in webpack.config.js to define env variables

[Deliver 159987629]

emmaadesile committed:
nitpicks(fix-travis-build): fix failing traivis build

[Delivers 159987629]

emmaadesile committed:
nitpicks(fix-merge-conflicts): fix merge conflicts

- rename article.js to createArticle.js

[Delivers 159987629]

emmaadesile committed:
nitpicks(create-article): update create article initial state

[Delievrs 159987629]

emmaadesile committed:
nitpicks(fix merge-conflicts): resolve merge conflicts from develop

[Delivers 159987629]

emmaadesile committed:
fix user create article feature

emekafredy reviewed (approved):
LGTM

CEOehis reviewed (approved):
LGTM

emmaadesile committed:
setup app for production

emmaadesile committed:
Merge branch 'develop' into ft-user-create-articles-159987629
